File: client/src/domain/models/Hand.ts
import Player, { PlayerCode } from '@/domain/models/Player';
import Card, { CardCode } from '@/domain/models/Card';

class Hand {
  public owner: Player;
  public cards: Card[];

  public constructor(owner: Player) {
    this.owner = owner;
    this.cards = new Array();
  }

  public isOwner(playerCode: PlayerCode) {
    return this.owner.code == playerCode;
  }

  public addCard(card: Card): void {
    this.cards.push(card);
  }

  public removeCard(cardToRemove: Card): void {
    const index = this.cards.findIndex((card) => card.equals(cardToRemove));
    this.cards.splice(index, 1);
  }

  public findCard(cardCode: CardCode): Card | undefined {
    return this.cards.find((card) => card.code.equals(cardCode));
  }
}

export default Hand;

File: src/main/java/com/dainws/games/crm/domain/core/player/Hand.java
package com.dainws.games.crm.domain.core.player;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

import com.dainws.games.crm.domain.core.card.Card;
import com.dainws.games.crm.domain.core.card.CardCode;
import com.dainws.games.crm.domain.core.card.CardType;
import com.dainws.games.crm.domain.core.exception.NotFoundException;

public class Hand {

	private List<Card> cards;

	public Hand() {
		this.cards = new ArrayList<>();
	}

	public void addCard(Card card) {
		this.cards.add(card);
	}

	public void addCards(List<Card> cards) {
		this.cards.addAll(cards);
	}

	public boolean contains(Card card) {
		return this.cards.contains(card);
	}

	public boolean contains(CardCode code) {
		return this.cards.stream().anyMatch(card -> card.getCode().equals(code));
	}

	public boolean contains(CardType type) {
		return this.cards.stream().anyMatch(card -> card.isType(type));
	}

	public void remove(Card card) {
		this.cards.remove(card);
	}

	public void remove(CardCode code) throws NotFoundException {
		Card card = this.getCard(code);
		this.cards.remove(card);
	}

	public void removeAllOf(CardType type) {
		this.cards.removeIf(card -> card.isType(type));
	}

	public Card grab(CardCode code) throws NotFoundException {
		Card card = this.getCard(code);
		this.cards.remove(card);
		return card;
	}

	public Card getCard(CardCode code) throws NotFoundException {
		Optional<Card> optionalCard = Optional.empty();

		for (Card card : this.cards) {
			if (card.getCode().equals(code)) {
				optionalCard = Optional.of(card);
			}
		}

		if (optionalCard.isEmpty()) {
			throw NotFoundException.cardNotFound();
		}

		return optionalCard.get();
	}

	public List<Card> getCardsOf(CardType type) {
		List<Card> searchedCards = new ArrayList<>();

		for (Card card : this.cards) {
			if (card.isType(type)) {
				searchedCards.add(card);
			}
		}

		return searchedCards;
	}

	public boolean isEmpty() {
		return this.cards.isEmpty();
	}
	
	public List<Card> getCards() {
		return cards;
	}
}
